### PR TITLE
Add ApiVersions

### PR DESCRIPTION
**Pull request by Pryz**
This PR adds the `ApiVersions` function to the API.

### CODE REVIEW

File: apiversions.go [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+	}
+
+	if remain, err = readInt32(r, size, &a.ThrottleTimeMs); err != nil {
+		return
+	}
+
+	return
+}
+
+type ApiVersionsV1 struct {
```
Comment (achille-roussel) [inline apiversions.go]: This is a protocol type, it shouldn't be exported.

File: conn.go [hunk 2]
```diff
@@ -989,6 +989,31 @@ func (c *Conn) requestHeader(apiKey apiKey, apiVersion apiVersion, correlationID
 	}
 }
 
+// ApiVersions return the list of API versions supported by the broker.
+func (c *Conn) ApiVersions() (versions []ApiVersionsV1, err error) {
```
Comment (achille-roussel) [inline conn.go]: Is there a reason why we would need to call this from a program? If not we shouldn't export it.

File: conn.go [hunk 3]
```diff
@@ -989,6 +989,31 @@ func (c *Conn) requestHeader(apiKey apiKey, apiVersion apiVersion, correlationID
...
+			versions = response.ApiVersions
+
+			return nil
+		},
+	)
+	if err != nil {
+		return []ApiVersionsV1{}, err
+	}
+
+	return
```
Comment (achille-roussel) [inline conn.go]: You're mixing names and unnamed return values, this is a bit confusing to the reader.

File: docker-compose.yml [hunk 4]
```diff
@@ -5,7 +5,7 @@ kafka:
   ports:
     - "9092:9092"
   environment:
-    KAFKA_VERSION: "0.10.1.0"
+    KAFKA_VERSION: "0.11.0.1"
```
Comment (achille-roussel) [inline docker-compose.yml]: I double-checked and it looks like this environment variable isn't used anymore, we should use an explicit tag on the image instead.

Comment (Pryz) [reply docker-compose.yml]: I think it's still in used : https://github.com/wurstmeister/kafka-docker/blob/master/Dockerfile#L8

File: conn.go (hunk 2 above)
Comment (Pryz) [reply conn.go]: I would say that it depends if the strategy is to let the user defining which protocol version to use or if we do that automatically in the library.

Comment (achille-roussel) [reply conn.go]: Let's do it automatically, kafka-go aims to be simple to use, this kind of implementation complexity is worth abstracting.

**Comment by Pryz** [comment]
I'm going to close this PR. The first idea was just to introduce the ability to use the ApiVersion API and it then became the "dealing with all the Kafka API versions".

Instead of doing that, we are going to rethink the internals of the library and use the latest message formats.

**Pull request closed without merging**